tests/Unit/Evolution/Systems/RelativisticEuler/Valencia/Test_Characteristics.cpp

```diff
@@ -127,16 +133,105 @@ void test_with_normal_along_coordinate_axes(
+                                 specific_enthalpy, equation_of_state);
+  const auto kappa_over_density = Scalar<double>{
+      {{get(equation_of_state
+                .kappa_times_p_over_rho_squared_from_density_and_energy(
+                    rest_mass_density, specific_internal_energy)) *
+        get(rest_mass_density) / get(pressure)}}};
+
+  const auto unit_normal_vector =
+      random_unit_normal(make_not_null(&generator), spatial_metric);
+  const auto unit_normal =
+      raise_or_lower_index(unit_normal_vector, spatial_metric);
+
+  Matrix right_matrix = RelativisticEuler::Valencia::right_eigenvectors(
+      rest_mass_density, spatial_velocity, specific_internal_energy, pressure,
+      specific_enthalpy, kappa_over_density, sound_speed_squared,
+      lorentz_factor, spatial_metric, det_and_inv_metric.second,
+      det_and_inv_metric.first, unit_normal);
+
+  Matrix left_matrix = RelativisticEuler::Valencia::left_eigenvectors(
+      rest_mass_density, spatial_velocity, specific_internal_energy, pressure,
+      specific_enthalpy, kappa_over_density, sound_speed_squared,
+      lorentz_factor, spatial_metric, det_and_inv_metric.second,
+      det_and_inv_metric.first, unit_normal);
+
+  const auto check_identity = [](const auto& left, const auto& right) noexcept {
+    Approx local_approx = Approx::custom().epsilon(1e-9).scale(1.0);
+    const Matrix left_times_right = left * right;
+    const Matrix right_times_left = right * left;
+    for (size_t i = 0; i < Dim + 2; ++i) {
+      for (size_t j = 0; j < Dim + 2; ++j) {
+        const double delta_ij = (i == j) ? 1. : 0.;
+        CHECK(left_times_right(i, j) == local_approx(delta_ij));
+        CHECK(right_times_left(i, j) == local_approx(delta_ij));
+      }
+    }
+  };
+  check_identity(left_matrix, right_matrix);
+
+  // test for unit normals along coordinate axes
+  for (const auto& direction : Direction<Dim>::all_directions()) {
+    auto basis_form = euclidean_basis_vector(direction, used_for_size);
+    const double inv_norm =
+        1.0 / get(magnitude(basis_form, det_and_inv_metric.second));
+    for (size_t i = 0; i < Dim; ++i) {
+      basis_form.get(i) *= inv_norm;
+    }
+    right_matrix = RelativisticEuler::Valencia::right_eigenvectors(
+        rest_mass_density, spatial_velocity, specific_internal_energy, pressure,
+        specific_enthalpy, kappa_over_density, sound_speed_squared,
+        lorentz_factor, spatial_metric, det_and_inv_metric.second,
+        det_and_inv_metric.first, basis_form);
+
+    left_matrix = RelativisticEuler::Valencia::left_eigenvectors(
+        rest_mass_density, spatial_velocity, specific_internal_energy, pressure,
+        specific_enthalpy, kappa_over_density, sound_speed_squared,
+        lorentz_factor, spatial_metric, det_and_inv_metric.second,
+        det_and_inv_metric.first, basis_form);
+
+    check_identity(left_matrix, right_matrix);
+  }
```

Comment:
Please add test that "right * speeds * left = d(flux)/du".

(Or is this far too difficult in the relativistic case? I am concerned that the current test does not check that these matrices are giving the correct transform... only that the two matrices are inverses of each other)

Reply:
I've worked out some terms and I was planning on adding that test too. The calculation is lengthy, so please do not wait if you need this code in develop.

Reply:
Okay, good to hear. I do not plan to use this transformation just yet (and can always get it from this branch), so I'm fine with waiting to merge this PR until you finish working it out. But if you prefer to do this as two steps with two PRs, that's okay with me too.